confdgnmi/src/confd_gnmi_api_adapter.py

```diff
@@ -142,52 +172,96 @@ def iter(kp, op, oldv, newv, state):
                     # TODO MODIFIED not handled for now
                 else:
                     log.warning(
-                        "Operation op=%d is not expected, kp=%s. Skipping!"
-                        , op, kp)
+                        "Operation op=%d is not expected, kp=%s. Skipping!",
+                        op, kp)
                 return _confd.ITER_RECURSE
 
-            cdb.diff_iterate(sub_sock, sub_point, iter, 0, None)
+            cdb.diff_iterate(sub_sock, sub_point, cdb_iter, 0, None)
             self.put_event(self.SubscriptionEvent.SEND_CHANGES)
             log.debug("self.change_db=%s", self.change_db)
             log.debug("<==")
 
-        def process_changes(self):
+        def process_external_change(self, ext_sock):
+            log.info("==>")
+            connection, client_address = ext_sock.accept()
+            with connection:
+                connection.setblocking(True)
+                # TODO make const
+                msg = connection.recv(1024)
+                log.debug("msg=%s", msg)
+                # simple protocol
+                # the msg string should contain N strings separated by \n
+                # op1\nxpath1\nval1\nop2\nxpath2\nval2 .....
+                # op1 .. first operation1, xpath1 .... first xpath, ...
+                # op is string used in ChangeOp Enum class
+                # currently operation can be only "modified"
+                # the size must be smaller then size in recv
+                data = msg.decode().split('\n')
+                assert len(data) % 3 == 0
+                chunks = [data[x:x + 3] for x in range(0, len(data), 3)]
+                self._append_changes(chunks)
+                self.put_event(self.SubscriptionEvent.SEND_CHANGES)
+                log.debug("data=%s", data)
+            log.info("<==")
+
+        def process_changes(self, external_changes=False):
```

Comment:
I'll reiterate my comment to your previous version, since the addition of external changes only made it worse: this method is quite long, it has 8 levels of nesting and is actually a method of a subclass. I would strongly suggest some refactoring here (such as: make a new class out of it and the two other `process_...` functions).

Reply:
Agree, but process_changes is still under 100 lines :-) An whole file under 1000 lines. I'll take a look at it during next iteration.